pointmatcher/DataPointsFilters/MaxPointCount.cpp

```diff
@@ -66,35 +67,20 @@ MaxPointCountDataPointsFilter<T>::filter(const DataPoints& input)
 template<typename T>
 void MaxPointCountDataPointsFilter<T>::inPlaceFilter(DataPoints& cloud)
 {
-	const size_t N = static_cast<size_t> (cloud.features.cols() - 1);
-	
-	if (maxCount <= N) 
+	const auto N = static_cast<size_t>(cloud.features.cols() - 1);
+
+	if (maxCount <= N)
 	{
-		//Re-init seed at each call, to ensure same results
-		std::srand(seed);
-		
-		for(size_t j=0; j<maxCount; ++j)
+		std::default_random_engine randomNumberGenerator(seed);
+
+		for(size_t j = 0; j < maxCount; ++j)
 		{
 			//Get a random index in [j; N]
-			const size_t idx = j + static_cast<size_t>((N-j)*(static_cast<float>(std::rand()/static_cast<float>(RAND_MAX))));
-			
-			//Switch columns j and idx
-			const auto feat = cloud.features.col(j);
-			cloud.features.col(j) = cloud.features.col(idx);
-			cloud.features.col(idx) = feat;
-			
-			if (cloud.descriptors.cols() > 0)
-			{
-				const auto desc = cloud.descriptors.col(j);
-				cloud.descriptors.col(j) = cloud.descriptors.col(idx);
-				cloud.descriptors.col(idx) = desc;
-			}
-			if (cloud.times.cols() > 0)
-			{
-				const auto time = cloud.times.col(j);
-				cloud.times.col(j) = cloud.times.col(idx);
-				cloud.times.col(idx) = time;
-			}	
+			std::uniform_int_distribution<size_t> distribution(j, N);
+			const size_t index = distribution(randomNumberGenerator);
+
+			//Switch columns j and index
+			cloud.swapCols(j, index);
```

Comment:
Can it happen that you swap a point that has already been swapped?

Reply:
I think this is a possibility, but I don't know if it can be a problem. To avoid this, we would need keep track of the points that has been already swapped. Do you think it can be a problem?